pkg/controller/persistentvolume/persistentvolume_claim_binder_controller.go

```diff
@@ -305,17 +326,41 @@ func syncClaim(volumeIndex *persistentVolumeOrderedIndex, binderClient binderCli
 			return fmt.Errorf("Unexpected error getting persistent volume: %v\n", err)
 		}
 
+		// TODO:  Can this be removed? The ref will not be nil for volumes bound by this controller.
 		if volume.Spec.ClaimRef == nil {
 			glog.V(5).Infof("Rebuilding bind on pv.Spec.ClaimRef\n")
 			claimRef, err := api.GetReference(claim)
 			if err != nil {
 				return fmt.Errorf("Unexpected error getting claim reference: %v\n", err)
 			}
-			volume.Spec.ClaimRef = claimRef
-			_, err = binderClient.UpdatePersistentVolume(volume)
+			clone, err := conversion.NewCloner().DeepCopy(volume)
 			if err != nil {
+				return fmt.Errorf("Error cloning pv: %v", err)
+			}
+			volumeClone, ok := clone.(*api.PersistentVolume)
+			if !ok {
+				return fmt.Errorf("Unexpected pv cast error : %v\n", volumeClone)
+			}
+			volumeClone.Spec.ClaimRef = claimRef
+			if updatedVolume, err := binderClient.UpdatePersistentVolume(volumeClone); err != nil {
 				return fmt.Errorf("Unexpected error saving PersistentVolume.Status: %+v", err)
+			} else {
+				volumeIndex.Update(updatedVolume)
+			}
+		}
+
+		// mismatched claims and bound volumes is an incorrect state.
+		// reset to Pending until api.ClaimError phase is introduced.
+		// the claim could become bound to another volume, which is appropriate.  This current volume
+		// does not belong to the claim.
+		if !claimAndVolumeMatch(claim, volume) {
```

Comment:
agree, this branch should be dropped, and the end of the ClaimPending branch should look like this:
```
		// the bind is persisted on the volume above and will always match the claim in a search.
		// claim would remain Pending if the update fails, but processing this state is idempotent.
		claim.Spec.VolumeName = volume.Name
		claim, err = binderClient.UpdatePersistentVolumeClaim(claim)
		if err != nil {
			return fmt.Errorf("Error updating claim with VolumeName %s: %+v\n", volume.Name, err)
		}

		claim.Status.Phase = api.ClaimBound
		claim.Status.AccessModes = volume.Spec.AccessModes
		claim.Status.Capacity = volume.Spec.Capacity
		_, err := binderClient.UpdatePersistentVolumeClaimStatus(claim)
		if err != nil {
			return fmt.Errorf("Unexpected error saving claim status: %+v", err)
		}
```